readthedocs/builds/version_slug.py

```diff
@@ -53,13 +54,15 @@ def get_fields_with_model(cls):
 
 class VersionSlugField(models.CharField):
 
-    """Inspired by ``django_extensions.db.fields.AutoSlugField``."""
+    """
+    Inspired by ``django_extensions.db.fields.AutoSlugField``.
 
-    invalid_chars_re = re.compile('[^-._a-z0-9]')
-    leading_punctuation_re = re.compile('^[-._]+')
-    placeholder = '-'
-    fallback_slug = 'unknown'
+    Uses ``unicode-slugify`` to generate the slug.
+    """
+
+    ok_chars = '-._'  # dash, dot, underscore
```

Comment:
From the rfc https://tools.ietf.org/html/rfc1035

```
The following syntax will result in fewer problems with many

applications that use domain names (e.g., mail, TELNET).

<domain> ::= <subdomain> | " "

<subdomain> ::= <label> | <subdomain> "." <label>

<label> ::= <letter> [ [ <ldh-str> ] <let-dig> ]

<ldh-str> ::= <let-dig-hyp> | <let-dig-hyp> <ldh-str>

<let-dig-hyp> ::= <let-dig> | "-"

<let-dig> ::= <letter> | <digit>
```

`_` underscore isn't allowed

Reply:
And it should start with a letter

Reply:
Well, I didn't change this behavior. We could consider.

Our current regex was working like that:

```
# Regex breakdown:
#   [a-z0-9] -- start with alphanumeric value
#   [-._a-z0-9] -- allow dash, dot, underscore, digit, lowercase ascii
#   *? -- allow multiple of those, but be not greedy about the matching
#   (?: ... ) -- wrap everything so that the pattern cannot escape when used in
#                regexes.
VERSION_SLUG_REGEX = '(?:[a-z0-9A-Z][-._a-z0-9A-Z]*?)'
```

Basically, it does not allow `.`, `-` and `_` as starting character, but it does allow a letter (lower and upper case) and a number.

If we want to be more strict and follow those directions, I suppose we should not allow the `.` (dot) either.

Reply:
Wait, this only affects the version's slug, not the project's slug. We are safe here, sorry for the noise.

Reply:
Shouldn't we use the same logic for project slugs if it's not explicitly defined by the user?

Reply:
Yeah, we use this for project's slug https://github.com/stsewd/readthedocs.org/blob/f1fd0c9827c10834277d63aa35cb4511ee3ae5ed/readthedocs/core/utils/__init__.py#L209-L219

Reply:
I don't mean removing the `_` and `.`. I mean, "why we don't use the all same logic for generating the slug for a Version than the slug for a Project?"

Reply:
Because Project's are used in DNS, but Versions aren't. They are different things, so we don't need the exact same restrictions.

Reply:
Yeah, I understand that.

I'm suggesting to use the same logic (unicode-slugify to generate the slug) but apply different restrictions.

Anyway, if so, it's for a different PR.